src/libs/Authentication.js

```diff
@@ -0,0 +1,148 @@
+                            throw new Error('passwordForm.error.incorrectLoginOrPassword');
+                        }
+                        throw new Error('passwordForm.error.twoFactorAuthenticationEnabled');
+                    case 403:
+                        if (response.message === 'Invalid code') {
+                            throw new Error('passwordForm.error.incorrect2fa');
+                        }
+                        throw new Error('passwordForm.error.invalidLoginOrPassword');
+                    case 404:
+                        throw new Error('passwordForm.error.unableToResetPassword');
+                    case 405:
+                        throw new Error('passwordForm.error.noAccess');
+                    case 413:
+                        throw new Error('passwordForm.error.accountLocked');
+                    default:
+                        throw new Error('passwordForm.error.fallback');
+                }
+            }
+            return response;
+        });
+}
+
+/**
+ * Reauthenticate using the stored credentials and redirect to the sign in page if unable to do so.
+ *
+ * @param {String} [command] command name for logging purposes
+ * @returns {Promise}
+ */
+function reauthenticate(command = '') {
+    // Prevent any more requests from being processed while authentication happens
+    NetworkStore.setIsAuthenticating(true);
+
+    const credentials = NetworkStore.getCredentials();
+    return Authenticate({
+        useExpensifyLogin: false,
+        partnerName: CONFIG.EXPENSIFY.PARTNER_NAME,
+        partnerPassword: CONFIG.EXPENSIFY.PARTNER_PASSWORD,
+        partnerUserID: credentials.autoGeneratedLogin,
+        partnerUserSecret: credentials.autoGeneratedPassword,
+    })
+        .then((response) => {
+            // If authentication fails throw so that we hit
+            // the catch below and redirect to sign in
+            if (response.jsonCode !== 200) {
+                throw new Error(response.message);
+            }
+
+            // Update authToken in Onyx and in our local variables so that API requests will use the
+            // new authToken
+            updateSessionAuthTokens(response.authToken, response.encryptedAuthToken);
+
+            // Note: It is important to manually set the authToken that is in the store here since any requests that are hooked into
+            // reauthenticate .then() will immediate post and use the local authToken. Onyx updates subscribers lately so it is not
+            // enough to do the updateSessionAuthTokens() call above.
+            NetworkStore.setAuthToken(response.authToken);
+
+            // The authentication process is finished so the network can be unpaused to continue
+            // processing requests
+            NetworkStore.setIsAuthenticating(false);
+        })
+
+        .catch((error) => {
+            // If authentication fails, then the network can be unpaused
+            NetworkStore.setIsAuthenticating(false);
```

Comment:
What happens if we fail authentication and unpause the network?  Does everything else just fail after with invalid authentication, should we also just cancel the rest of the requests here?

Reply:
It depends on why it failed and which queue is processing so we have to look at the `Retry` middleware to figure it out.

If we are in the main queue and this failed because of a networking issue then requests will get put back into the queue and will either run again when we come back online (if we went offline) or if we are online they will run again until we run out of retries. **Note:** That behavior will change soon because we will soon stop retrying basically "all requests" like we do now.

https://github.com/Expensify/App/blob/f1c8b4fa4570cee4820c079d8276243e8a350659/src/libs/Middleware/Retry.js#L58-L68

If we are in the sequential queue then we will leave the request in the queue but move onto the next one.

https://github.com/Expensify/App/blob/f72c7d510e522587dffcdc72464c153d1d7d69e2/src/libs/Network/SequentialQueue.js#L28-L33

On the other hand if this fails because we couldn't authenticate at all (and there are no networking issues) then the user must be logged out.

And the act of logging out will cancel the requests in the main queue and clear storage (which I _think_ would also take care of any persisted requests that haven't run yet).

https://github.com/Expensify/App/blob/f72c7d510e522587dffcdc72464c153d1d7d69e2/src/libs/actions/SignInRedirect.js#L50-L52

Reply:
Ah gotcha, sweet thanks!